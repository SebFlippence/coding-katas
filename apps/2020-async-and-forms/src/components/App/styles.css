html, body {
  height: 100%;
  margin: 0;
  font-family: sans-serif;
}

#root {
  height: 100%;
}

.exercise-list__item {
  display: block;
}
