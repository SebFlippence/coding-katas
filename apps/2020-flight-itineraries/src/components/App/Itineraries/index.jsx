import React from 'react';
import BpkCard from 'bpk-component-card';
import BpkButton from 'bpk-component-button';
import {
  BpkGridContainer,
  BpkGridRow,
  BpkGridColumn,
} from 'bpk-component-grid';

import STYLES from './index.scss';
import Legs from './Legs';

const Itinerary = ({ item }) => (
  <BpkGridContainer fullWidth>
    <BpkGridRow>
      <BpkGridColumn width={8}>
        <span className={STYLES.price}>{item.price}</span>
        <br />
        <span className={STYLES.agent}>{item.agent}</span>
      </BpkGridColumn>
      <BpkGridColumn width={1}>
        <BpkButton>Select</BpkButton>
      </BpkGridColumn>
    </BpkGridRow>
  </BpkGridContainer>
);

const Itineraries = ({ flights }) =>
  flights.itineraries.map(itinerary => (
    <section key={itinerary.id} className={STYLES.item}>
      <BpkCard className={STYLES.card}>
        <Legs
          className="Legs"
          data={flights.legs.filter(leg => itinerary.legs.includes(leg.id))}
        />
        <Itinerary item={itinerary} />
      </BpkCard>
    </section>
  ));

export default Itineraries;
