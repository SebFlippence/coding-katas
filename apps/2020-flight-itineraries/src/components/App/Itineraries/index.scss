.card {
  width: 365px;
  margin-left: 10px;
}

.item {
  margin-bottom: 20px;
}

.price {
  font-size: 1.5rem;
}

.agent {
  color: #ccc;
}
