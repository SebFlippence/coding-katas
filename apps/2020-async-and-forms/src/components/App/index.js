import React from "react";

import Complexity from "../Complexity";
import Form from "../Form";

import "./styles.css";

export default () => {
  const page = window.location.pathname.slice(1);

  return (
    <>
      {page.length === 0 && (
        <ul className="exercise-list">
          <li className="exercise-list__item">
            <a href="/complexity">Complexity</a>
          </li>
          <li className="exercise-list__item">
            <a href="/form">Form</a>
          </li>
        </ul>
      )}

      {page === "complexity" && <Complexity />}

      {page === "form" && <Form />}
    </>
  );
};
