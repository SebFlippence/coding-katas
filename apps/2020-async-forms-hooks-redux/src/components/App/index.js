import React from 'react';

import Complexity from '../Complexity';
import Form from '../Form';
import FormHooks from '../FormHooks';
import CustomHook from '../CustomHook';
import ReduxToolkit from '../ReduxToolkit';

import './styles.css';

export default () => {
  const page = window.location.pathname.slice(1);

  return (
    <>
      {page.length === 0 && (
        <ul className="exercise-list">
          <li className="exercise-list__item">
            <a href="/complexity">Complexity</a>
          </li>
          <li className="exercise-list__item">
            <a href="/form">Form</a>
          </li>
          <li className="exercise-list__item">
            <a href="/form-hooks">React Form Hook</a>
          </li>
          <li className="exercise-list__item">
            <a href="/custom-hook">Custom Hook</a>
          </li>
          <li className="exercise-list__item">
            <a href="/redux-toolkit">Redux Toolkit</a>
          </li>
        </ul>
      )}

      {page === 'form' && <Form />}

      {page === 'form-hooks' && <FormHooks />}

      {page === 'custom-hook' && <CustomHook />}

      <div style={{ margin: 10 }}>
        {page === 'complexity' && <Complexity />}

        {page === 'redux-toolkit' && <ReduxToolkit />}
      </div>
    </>
  );
};
