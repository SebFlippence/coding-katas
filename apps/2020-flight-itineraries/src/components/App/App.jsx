import React, { useEffect, useState } from 'react';
import BpkText from 'bpk-component-text';

import Header from '../Header';

import STYLES from './App.scss';
import Itineraries from './Itineraries';

const getClassName = className => STYLES[className] || 'UNKNOWN';

const App = () => {
  const [flights, setFlights] = useState();

  useEffect(() => {
    async function fetchData() {
      const response = await fetch('flights.json');
      if (!response.ok) {
        throw new Error('There was a problem loading flights');
      }
      setFlights(await response.json());
    }
    fetchData();
  }, []);

  return (
    <div className={getClassName('App')}>
      <Header />
      <main className={getClassName('App__main')}>
        {flights ? (
          <Itineraries flights={flights} />
        ) : (
          <BpkText tagName="p">Looking for flights...</BpkText>
        )}
      </main>
    </div>
  );
};

export default App;
