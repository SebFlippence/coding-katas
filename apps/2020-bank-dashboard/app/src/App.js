import React, { useEffect } from 'react';
import { observer, inject } from 'mobx-react';
import './App.css';

function App({ bankStore }) {
  useEffect(() => {
    bankStore.getTransactions();
  }, [bankStore]);

  return (
    <div className="App">
      <h1>Transactions</h1>
      {!bankStore.transactions.length ? (
        <p>Loading...</p>
      ) : (
        <table>
          <thead>
            <tr>
              <th>Amount</th>
              <th>User</th>
              <th>Retailer</th>
            </tr>
          </thead>
          <tbody>
            {bankStore.transactions.map(transaction => (
              <tr key={transaction.id}>
                <td>
                  {transaction.amount} {transaction.currency}
                </td>
                <td>{bankStore.users[transaction.userId] ? bankStore.users[transaction.userId].name : 'Loading...'}</td>
                <td>
                  {bankStore.retailers[transaction.retailerId]
                    ? `${bankStore.retailers[transaction.retailerId].name} ${
                        bankStore.retailers[transaction.retailerId].city
                      }`
                    : 'Loading...'}
                </td>
              </tr>
            ))}
          </tbody>
        </table>
      )}
    </div>
  );
}

export default inject(({ bankStore }) => ({ bankStore }))(observer(App));
