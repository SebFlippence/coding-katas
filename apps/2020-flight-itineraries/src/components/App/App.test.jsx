import React from 'react';
import ReactDOM from 'react-dom';
import { render } from '@testing-library/react';
import { act } from 'react-test-renderer';
import { waitForElementToBeRemoved } from '@testing-library/dom';

import App from './App';

describe('App', () => {
  beforeEach(() => {
    global.fetch = jest.fn().mockImplementation(
      () =>
        new Promise(resolve =>
          resolve({
            ok: true,
            json() {
              return { itineraries: [], legs: [] };
            },
          }),
        ),
    );
  });

  it('should render without crashing', async () => {
    const div = document.createElement('div');

    ReactDOM.render(<App />, div);
  });

  it('should render correctly while loading', async () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with flight results', async () => {
    await act(async () => {
      const { getByText, container } = render(<App />);
      await waitForElementToBeRemoved(() =>
        getByText('Looking for flights...'),
      );

      expect(container).toMatchSnapshot();
    });

    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith('flights.json');
  });

  // @TODO investigate testing error conditions
  it.skip('should show an error if flights fail to load', async () => {
    global.fetch = jest.fn().mockImplementation(
      () =>
        new Promise(resolve =>
          resolve({
            ok: false,
            json() {
              return { error: {} };
            },
          }),
        ),
    );

    expect(async () => {
      await act(async () => {
        const { getByText } = render(<App />);
        await waitForElementToBeRemoved(() =>
          getByText('Looking for flights...'),
        );

        expect(global.fetch).toHaveBeenCalledTimes(1);
        expect(global.fetch).toHaveBeenCalledWith('flights.json');
      });
    }).toThrowError('There was a problem loading flights');
  });
});
