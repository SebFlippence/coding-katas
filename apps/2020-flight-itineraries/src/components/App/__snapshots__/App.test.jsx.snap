// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly while loading 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="Header"
    >
      <a
        href="/"
      >
        <span
          class="Header__hidden-text"
        >
          Flight Itineraries
        </span>
        <img
          alt="Flight Itineraries"
          class="Header__logo-image"
          src="logo192.png"
        />
      </a>
    </header>
    <main
      class="App__main"
    >
      <p
        class="bpk-text bpk-text--base"
      >
        Looking for flights...
      </p>
    </main>
  </div>
</div>
`;

exports[`App should render correctly with flight results 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="Header"
    >
      <a
        href="/"
      >
        <span
          class="Header__hidden-text"
        >
          Flight Itineraries
        </span>
        <img
          alt="Flight Itineraries"
          class="Header__logo-image"
          src="logo192.png"
        />
      </a>
    </header>
    <main
      class="App__main"
    />
  </div>
</div>
`;

exports[`App should show an error if flights fail to load 1`] = `
<div>
  <div
    class="App"
  >
    <header
      class="Header"
    >
      <a
        href="/"
      >
        <span
          class="Header__hidden-text"
        >
          Skyscanner
        </span>
        <img
          alt="Skyscanner"
          class="Header__logo-image"
          src="logo.svg"
        />
      </a>
    </header>
    <main
      class="App__main"
    >
      <h2
        class="bpk-text bpk-text--base"
      >
        Your flight results
      </h2>
      hello
    </main>
  </div>
</div>
`;
